src/plugins/kibana_utils/docs/state_sync/on_fly_state_migrations.md

```diff
@@ -0,0 +1,46 @@
+# On-fly state migrations
+
+When retrieving initial state from storage we shouldn't forget about possible outdated state.
```

Comment:
I think the explanation about the scenario should come first - 

Imagine an edge case scenario, where a user is working in your application, and then pastes an old link for the same application, containing older state with a different structure.
Since no application remount will happen, we need to transition to a new state on-the-fly, by applying necessary migrations.